pkg/util/kube/storage.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package kube
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	pkgcfg "github.com/vmware-tanzu/vm-operator/pkg/config"
+	"github.com/vmware-tanzu/vm-operator/pkg/util/kube/internal"
+)
+
+// GetStoragePolicyID returns the storage policy ID for a given StorageClass.
+// If no ID is found, an error is returned.
+func GetStoragePolicyID(obj storagev1.StorageClass) (string, error) {
+	policyID, ok := obj.Parameters[internal.StoragePolicyIDParameter]
+	if !ok {
+		return "", fmt.Errorf(
+			"StorageClass %q does not have '"+
+				internal.StoragePolicyIDParameter+"' parameter",
+			obj.Name)
+	}
+	return policyID, nil
+}
+
+// MarkEncryptedStorageClass records the provided StorageClass as encrypted.
+func MarkEncryptedStorageClass(
+	ctx context.Context,
+	k8sClient ctrlclient.Client,
+	storageClass storagev1.StorageClass,
+	encrypted bool) error {
+
+	var (
+		obj = corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: pkgcfg.FromContext(ctx).PodNamespace,
+				Name:      internal.EncryptedStorageClassNamesConfigMapName,
+			},
+		}
+		ownerRef = internal.GetOwnerRefForStorageClass(storageClass)
+	)
+
+	// Get the ConfigMap.
+	err := k8sClient.Get(ctx, ctrlclient.ObjectKeyFromObject(&obj), &obj)
+
+	if err != nil {
+		// Return any error other than 404.
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+
+		//
+		// The ConfigMap was not found.
+		//
+
+		if !encrypted {
+			// If the goal is to mark the StorageClass ss not encrypted, then we
+			// do not need to actually create the underlying ConfigMap if it does
+			// not exist.
+			return nil
+		}
+
+		// The ConfigMap does not already exist and the goal is to mark the
+		// StorageClass as encrypted, so go ahead and create the ConfigMap and
+		// return early.
+		obj.OwnerReferences = []metav1.OwnerReference{ownerRef}
```

Comment:
I am using OwnerRefs because Sandeep will not allow me to annotate the StorageClasses and he did not like me placing a finalizer on them either when I was directly modifying the ConfigMap with the SC names. Neither the GoDocs nor the Kube docs mention any limit.